roles/matrix-postgres/tasks/import_generic_sqlite_db.yml

```diff
@@ -0,0 +1,84 @@
+---
+
+# Pre-checks
+
+- name: Fail if Postgres not enabled
+  fail:
+    msg: "Postgres via the matrix-postgres role is not enabled (`matrix_postgres_enabled`). Cannot import."
+  when: "not matrix_postgres_enabled|bool"
+
+- name: Fail if playbook called incorrectly
+  fail:
+    msg: "The `sqlite_database_path` variable needs to be provided to this playbook, via --extra-vars"
+  when: "sqlite_database_path is not defined or sqlite_database_path.startswith('<')"
+
+- name: Fail if playbook called incorrectly
+  fail:
+    msg: >-
+      The `postgres_db_connection_string` variable needs to be provided to this playbook, via `--extra-vars`.
+      Example: `--extra-vars="postgres_db_connection_string=postgresql://username:password@localhost:<port>/database_name`"
+  when: "postgres_db_connection_string is not defined or not postgres_db_connection_string.startswith('postgresql://')"
+
+- name: Check if the provided SQLite database file exists
+  stat:
+    path: "{{ sqlite_database_path }}"
+  register: sqlite_database_path_stat_result
+
+- name: Fail if provided SQLite database file doesn't exist
+  fail:
+    msg: "File cannot be found on the server at {{ sqlite_database_path }}"
+  when: "not sqlite_database_path_stat_result.stat.exists"
+
+
+# Defaults
+
+- name: Set postgres_start_wait_time, if not provided
+  set_fact:
+    postgres_start_wait_time: 15
+  when: "postgres_start_wait_time|default('') == ''"
+
+
+# Actual import work
+
+- name: Ensure matrix-postgres is started
+  service:
+    name: matrix-postgres
+    state: started
+    daemon_reload: yes
+  register: matrix_postgres_service_start_result
+
+- name: Wait a bit, so that Postgres can start
+  wait_for:
```

Comment:
There is a feature somewhere which can do stuff like `wait_for` a specific port to be available:
```
- wait_for:
    port: 5432
    delay: 5
    timeout: "{{ postgres_start_wait_time }}"
```
(though this requires getting the IP of `matrix-postgres, but that's not hard)

Reply:
We can extract it, and I was initially redoing your extraction logic to not require `jq`, etc., but I still don't think that we can reach that IP from the host itself.

Has that worked for you somehow?

I'd think that a container binding to ports on some network (well, any network besides `--net=host`) will not be reachable from the host itself (or from any other container network), unless ports have been exposed (`-p 5432:5432`) or some other kind of bridging is configured.

-----------------

What we could do is use something like [waitforit](https://github.com/maxcnunes/waitforit) from within a container on the same network (`--network=matrix`) and wait for `matrix-postgres:5432`. We'd need to find a container image for it though. And it'd be nice it it worked on ARM architectures as well.

Having said that, our current pgloader container image (`docker.io/illagrenan/pgloader:3.6.2`) only works on `amd64`, so people on ARM don't have a good alternative :/ Still, that's something that only affects people on current installations, who want to move to Postgres.

But if we are to do this `waitforit` thing, it would execute for everyone who enables at least one service (bridge) that uses an additional database.

Maybe waiting for 10-15 seconds is not too terrible, given that it's easy to do and doesn't break ARM support.

Reply:
Something like this has worked for me in the past. we can `docker inspect` the container, register the output and use `|from_json` and lookup the IP, and on all docker daemon variants, the IP itself should always be routable from the host to the container.

If the pg_loader is `amd64` only right now, `arm64` people wont be able to migrate anyway

Reply:
That's an easy way to get it without needing `jq`: `docker inspect matrix-postgres --format '{{.NetworkSettings.Networks.matrix.IPAddress}}'`

It yields something like `172.18.0.3` for me right now.

Connecting doesn't work for me though (testing `telnet 172.18.0.3 5432` from the host).

And I'd be surprised if it did, as that's the whole point of containers -- isolating these processes and their network away from the host (and any other containers), unless specifically requested to bridge.

This does work though: `docker run -it --rm --network=matrix docker.io/alpine:3.12 /bin/sh -c 'apk add --no-cache busybox-extras && telnet 172.18.0.3 5432'`

Reply:
As for the ARM thing, yes, ARM users won't be able to migrate anyway as things are right now. But at least new playbook users can still do a full installation (which uses bridges with Postgres) without any breakage. If run `waitforit` in some `amd64`-only container, we'd be breaking `--tags=setup-all` for anyone who has enabled anything that uses additional Postgres databases, unless we find or build our own ARM image.

I feel it's better if we don't break the main installation path and rather just break some migration command for some subset of the users. For now, at least. Not that these are related. They're both fixable, separately ;) And we should